README.md
parallelTools
=============

```
library(devtools)
install_github("wahani/parallelTools", subdir = "package")
```


package/NEWS
Version 0.2
- global setting for packageToLoad and sourceFile instead of function arguments
- renaming of mcapply from Version 0.1 to mclapply.0.1

Version 0.1
- parallelTools::mclapply is a simple wrapper function for parallel::mclapply when called in linux. In Windows the function will use parallel::clusterApply/parallel::clusterApplyLB.

R_Code/test.R
if(.Platform$OS.type == "windows") {
  require(devtools)
  install_github(repo="parallelTools", username = "wahani", subdir = "package")
}
require(parallelTools)

genData <- function(n) {
  # Generate some Data
  dat <- data.frame()
  for (i in 1:n) {
    dat <- rbind(dat, data.frame(colA = letters[sample(26)], colB = letters[sample(26)]))
  }
  dat
}

slowFunction <- function(dat) {
  # A slow function
  dat$ind <- logical(nrow(dat))
  
  for (i in 1:nrow(dat)) {
    dat$ind[i] <- grepl(dat$colA[i], dat$colB[i])
  }
  
  dat
}

nCores <- detectCores() - 1 # Detects the number of cores
dat <- genData(1000) # generate some Date
dat$sample <- rep_len(1:nCores, length.out = nrow(dat)) # ID for splitting the data
dat$id <- 1:nrow(dat) # ID for observations
dataList <- split(dat, dat$sample) # split the dataSet for processing the tasks parallel

system.time(result1 <- slowFunction(dat)) # Function without mc

# Function with mc - should work for all platforms
system.time(result2 <- do.call("rbind", mclapply(X = dataList, FUN = slowFunction, mc.cores = nCores)))

# Better than mc is vecorization (in this case) - no multicore required:
system.time(result3 <- mapply(grepl, as.list(dat$colA), as.list(dat$colB)))

#All the same
all((result1$ind == result2$ind[order(result2$id)]) & result1$ind == result3)

R_Code/test1.R
# getPTOption <- function() {
#   ptOptions <- .Options[grepl("parallelTools", names(.Options))]
#   if(length(ptOptions) == 0) {
#     cat("No parrallelTools options set\n")
#     invisible(NULL)
#   } else {
#     cat("packages that will be loaded on each cluster:", 
#         if(ptOptions$parallelToolsPTL == "") "\"nothing\"" else ptOptions$parallelToolsPTL,
#         "\nsource files sourced on each cluster:", 
#         if(ptOptions$parallelToolsSF == "") "\"nothing\"" else ptOptions$parallelToolsSF)
#     invisible(ptOptions)
#   }
# }
# 
# setPTOption <- function(packageToLoad = "", sourceFile = "") {
#   options(parallelToolsPTL = packageToLoad, parallelToolsSF = sourceFile)
# }

setPTOption("testthat")
getPTOption()








genData <- function(n) {
  # Generate some Data
  dat <- data.frame()
  for (i in 1:n) {
    dat <- rbind(dat, data.frame(colA = letters[sample(26)], colB = letters[sample(26)]))
  }
  dat
}

slowFunction <- function(dat) {
  # A slow function
  dat$ind <- logical(nrow(dat))
  
  for (i in 1:nrow(dat)) {
    dat$ind[i] <- grepl(dat$colA[i], dat$colB[i])
  }
  
  dat
}

nCores <- detectCores() - 1 # Detects the number of cores
dat <- genData(1000) # generate some Date
dat$sample <- rep_len(1:nCores, length.out = nrow(dat)) # ID for splitting the data
dat$id <- 1:nrow(dat) # ID for observations
dataList <- split(dat, dat$sample) # split the dataSet for processing the tasks parallel

system.time(result1 <- slowFunction(dat)) # Function without mc

# Function with mc - should work for all platforms
system.time(result2 <- do.call("rbind", mclapply(X = dataList, FUN = slowFunction, mc.cores = nCores)))

# Better than mc is vecorization (in this case) - no multicore required:
system.time(result3 <- mapply(grepl, as.list(dat$colA), as.list(dat$colB)))

#All the same
all((result1$ind == result2$ind[order(result2$id)]) & result1$ind == result3)


R_Code/MClapply.R
# MClapply <- function(X, FUN, ..., mc.preschedule = TRUE, mc.set.seed = TRUE,
#                      mc.silent = FALSE, mc.cores = 1L,
#                      mc.cleanup = TRUE, mc.allow.recursive = TRUE, 
#                      packageToLoad = "", sourceFile = "") {
#   require(parallel)
#   if(.Platform$OS.type == "windows") {
#     cl <- makeCluster(mc.cores)
#     clusterEvalQ(cl, expression(lapply(packageToLoad, library)))
#     clusterFunction <- if(mc.preschedule) clusterApply else clusterApplyLB
#     return(clusterFunction(cl = cl, x = X, fun = FUN, ...=...))
#     
#   } else {
#     return(mclapply(X = X, FUN = FUN, ...=..., mc.preschedule = mc.preschedule, 
#                     mc.set.seed = mc.set.seed, mc.silent = mc.silent, mc.cores = mc.cores,
#                     mc.cleanup = mc.cleanup, mc.allow.recursive = mc.allow.recursive))
#   }
# }

# mclapply(lapply(10^1:20, rnorm), function(x) {
#   Sys.sleep(10)
#   mean(x)
# }, mc.cores = 2)

1:10


package/R/mclapply.R
#' mclapply - Wrapper for parallel::mclapply/clusterApply
#' 
#' @description This function masks the function mclapply from package parallel. The aim is
#' to use multicore functionality under windows and linux. For the description of the function
#' arguments see: \link[parallel]{mclapply} in the package parallel
#' 
#' @inheritParams parallel::mclapply
#' @param mc.preschedule see \code{\link[parallel]{mclapply}} for Linux. Windows: If TRUE \code{\link{clusterApply}}
#' will be used, if FALSE \code{\link{clusterApplyLB}} is used
#' @param packageToLoad these two arguments are only relevant using Windows. packageToLoad is a 
#' character vector with the names of the packages which should be loaded using \code{\link{require}} on to each
#' cluster. In addition with sourceFile a R-Script can be specified which will be sourced on each cluster.
#' @param sourceFile see \code{packageToLoad}
#' @param mc.cores number of cores to be used.
#' 
#' @details The function will detect th operating system via \code{.Platform$OS.type}. If the platform is equal
#' to "windows", \code{mc.cores} clusters will be initialized using \code{\link{makeCluster}}. Packages and sourcefiles
#' will be loaded onto each cluster using \code{\link{clusterEvalQ}}. Then either \code{\link{clusterApply}}
#' or \code{\link{clusterApplyLB}} is called with the arguments \code{x = X, fun = FUN, ... = ...} - \code{mc.set.seed},
#' \code{mc.silent}, \code{mc.cleanup} and \code{mc.allow.recursive} will be ignored under windows. The function will end with 
#' \code{\link{stopCluster}} to end all started sessions and return the resulting list.
#' 
#' If \code{.Platform$OS.type} is not equal to windows the function \code{\link[parallel]{mclapply}} will be called, passing
#' all arguments but \code{packageToLoad} and \code{sourceFile} which will be ignored.
#' 
#' This function is written to use multicore functionality in developing mainly for linux but with the ability to test
#' the code in windows. The side effects of the applied parallelization for windows I do not know. Check the running 
#' processes regularly.
#' 
#' @seealso Everything in the package "parallel": \link{clusterApply} and \link[parallel]{mclapply} and \link{mcmapply}
#' @author Sebastian Warnholz
#' @examples 
#' genData <- function(n) {
#'  # Generate some Data
#'  dat <- data.frame()
#'  for (i in 1:n) {
#'    dat <- rbind(dat, data.frame(colA = letters[sample(26)], colB = letters[sample(26)]))
#'  }
#'  dat
#' }
#'
#' slowFunction <- function(dat) {
#'  # A slow function
#'  dat$ind <- logical(nrow(dat))
#'  
#'  for (i in 1:nrow(dat)) {
#'    dat$ind[i] <- grepl(dat$colA[i], dat$colB[i])
#'  }
#'  
#'  dat
#' }
#'
#' nCores <- detectCores() - 1 # Detects the number of cores
#' dat <- genData(1000) # generate some Date
#' dat$sample <- rep_len(1:nCores, length.out = nrow(dat)) # ID for splitting the data
#' dat$id <- 1:nrow(dat) # ID for observations
#' dataList <- split(dat, dat$sample) # split the dataSet for processing the tasks parallel
#'
#' system.time(result1 <- slowFunction(dat)) # Function without mc
#'
#' # Function with mc - should work for all platforms
#' system.time(result2 <- do.call("rbind", mclapply(X = dataList, FUN = slowFunction, mc.cores = nCores)))
#'
#' # Better than mc is vecorization (in this case) - no multicore required:
#' system.time(result3 <- mapply(grepl, as.list(dat$colA), as.list(dat$colB)))
#'
#' #All the same
#' all((result1$ind == result2$ind[order(result2$id)]) & result1$ind == result3)
#' @export
mclapply <- function(X, FUN, ..., mc.preschedule = TRUE, mc.set.seed = TRUE,
                     mc.silent = FALSE, mc.cores = 1L,
                     mc.cleanup = TRUE, mc.allow.recursive = TRUE, 
                     packageToLoad = "", sourceFile = "") {
  if(.Platform$OS.type == "windows") {
    clusterFunction <- if(mc.preschedule) clusterApply else clusterApplyLB
    cl <- makeCluster(mc.cores)
    if(all(packageToLoad != "")) {
      clusterExport(cl, "packageToLoad", envir = environment())
      clusterEvalQ(cl, lapply(packageToLoad, require, character.only = TRUE))
    }
    if(sourceFile != "") {
      clusterExport(cl, "sourceFile", envir = environment())
      clusterEvalQ(cl, source(sourceFile, echo=FALSE))
    }
    result <- clusterFunction(cl = cl, x = X, fun = FUN, ...=...)
    stopCluster(cl)
    return(result)
  } else {
    return(parallel::mclapply(X = X, FUN = FUN, ...=..., mc.preschedule = mc.preschedule, 
                              mc.set.seed = mc.set.seed, mc.silent = mc.silent, mc.cores = mc.cores,
                              mc.cleanup = mc.cleanup, mc.allow.recursive = mc.allow.recursive))
  }
}

package/R/getSetPTOptions.R
#' get/set packages/sourceFiles to be loaded for Windows
#' @description get and set the packages and source files which will be loaded on to each cluster. 
#' Only relevant in Windows when the function applied in \code{\link[parallelTools]{mclapply}} depends
#' on some package or objects which can be sourced.
#' 
#' @inheritParams parallelTools::mclapply.0.1
#' 
#' @seealso \link[parallelTools]{mclapply}
#' @author Sebastian Warnholz
#' @rdname getSetPTOOptions
#' @export
getPTOption <- function() {
  ptOptions <- .Options[grepl("parallelTools", names(.Options))]
  if(length(ptOptions) == 0) {
    cat("No parrallelTools options set\n")
    invisible(NULL)
  } else {
    cat("packages that will be loaded on each cluster:", 
        if(ptOptions$parallelToolsPTL == "") "\"nothing\"" else ptOptions$parallelToolsPTL,
        "\nsource files sourced on each cluster:", 
        if(ptOptions$parallelToolsSF == "") "\"nothing\"" else ptOptions$parallelToolsSF)
    invisible(ptOptions)
  }
}

#' @rdname getSetPTOOptions
#' @export
setPTOption <- function(packageToLoad = "", sourceFile = "") {
  options(parallelToolsPTL = packageToLoad, parallelToolsSF = sourceFile)
}


parallelTools.Rcheck/parallelTools-Ex.R
pkgname <- "parallelTools"
source(file.path(R.home("share"), "R", "examples-header.R"))
options(warn = 1)
options(pager = "console")
library('parallelTools')

base::assign(".oldSearch", base::search(), pos = 'CheckExEnv')
cleanEx()
nameEx("mclapply")
### * mclapply

flush(stderr()); flush(stdout())

### Name: mclapply
### Title: mclapply - Wrapper for parallel::mclapply/clusterApply
### Aliases: mclapply

### ** Examples

genData <- function(n) {
  dat <- data.frame()
  for (i in 1:n) {
    dat <- rbind(dat, data.frame(colA = letters[sample(26)], colB = letters[sample(26)]))
  }
  dat
}

slowFunction <- function(dat) {
  dat$ind <- logical(nrow(dat))

  for (i in 1:nrow(dat)) {
    dat$ind[i] <- grepl(dat$colA[i], dat$colB[i])
  }

  dat
}

nCores <- detectCores() - 1
dat <- genData(1000)
dat$sample <- gl(nCores, 1)
dataList <- split(dat, dat$sample)

system.time(result1 <- slowFunction(dat))
system.time(result2 <- mclapply(X = dataList, FUN = slowFunction, mc.cores = nCores))



### * <FOOTER>
###
options(digits = 7L)
base::cat("Time elapsed: ", proc.time() - base::get("ptime", pos = 'CheckExEnv'),"\n")
grDevices::dev.off()
###
### Local variables: ***
### mode: outline-minor ***
### outline-regexp: "\\(> \\)?### [*]+" ***
### End: ***
quit('no')


package/R/mclapply.0.1.R
#' mclapply.0.1
#' 
#' @description Old version of parallelTools::mclapply. This function masks the function mclapply from package parallel. The aim is
#' to use multicore functionality under windows and linux. For the description of the function
#' arguments see: \link[parallel]{mclapply} in the package parallel
#' 
#' @inheritParams parallel::mclapply
#' @param mc.preschedule see \code{\link[parallel]{mclapply}} for Linux. Windows: If TRUE \code{\link{clusterApply}}
#' will be used, if FALSE \code{\link{clusterApplyLB}} is used
#' @param packageToLoad these two arguments are only relevant using Windows. packageToLoad is a 
#' character vector with the names of the packages which should be loaded using \code{\link{require}} on to each
#' cluster. In addition with sourceFile a R-Script can be specified which will be sourced on each cluster.
#' @param sourceFile see \code{packageToLoad}
#' @param mc.cores number of cores to be used.
#' 
#' @details The function will detect th operating system via \code{.Platform$OS.type}. If the platform is equal
#' to "windows", \code{mc.cores} clusters will be initialized using \code{\link{makeCluster}}. Packages and sourcefiles
#' will be loaded onto each cluster using \code{\link{clusterEvalQ}}. Then either \code{\link{clusterApply}}
#' or \code{\link{clusterApplyLB}} is called with the arguments \code{x = X, fun = FUN, ... = ...} - \code{mc.set.seed},
#' \code{mc.silent}, \code{mc.cleanup} and \code{mc.allow.recursive} will be ignored under windows. The function will end with 
#' \code{\link{stopCluster}} to end all started sessions and return the resulting list.
#' 
#' If \code{.Platform$OS.type} is not equal to windows the function \code{\link[parallel]{mclapply}} will be called, passing
#' all arguments but \code{packageToLoad} and \code{sourceFile} which will be ignored.
#' 
#' This function is written to use multicore functionality in developing mainly for linux but with the ability to test
#' the code in windows. The side effects of the applied parallelization for windows I do not know. Check the running 
#' processes regularly.
#' 
#' @seealso Everything in the package "parallel": \link{clusterApply} and \link[parallel]{mclapply} and \link{mcmapply}
#' @author Sebastian Warnholz
#' @examples 
#' genData <- function(n) {
#'  # Generate some Data
#'  dat <- data.frame()
#'  for (i in 1:n) {
#'    dat <- rbind(dat, data.frame(colA = letters[sample(26)], colB = letters[sample(26)]))
#'  }
#'  dat
#' }
#'
#' slowFunction <- function(dat) {
#'  # A slow function
#'  dat$ind <- logical(nrow(dat))
#'  
#'  for (i in 1:nrow(dat)) {
#'    dat$ind[i] <- grepl(dat$colA[i], dat$colB[i])
#'  }
#'  
#'  dat
#' }
#'
#' nCores <- detectCores() - 1 # Detects the number of cores
#' dat <- genData(1000) # generate some Date
#' dat$sample <- rep_len(1:nCores, length.out = nrow(dat)) # ID for splitting the data
#' dat$id <- 1:nrow(dat) # ID for observations
#' dataList <- split(dat, dat$sample) # split the dataSet for processing the tasks parallel
#'
#' system.time(result1 <- slowFunction(dat)) # Function without mc
#'
#' # Function with mc - should work for all platforms
#' system.time(result2 <- do.call("rbind", mclapply(X = dataList, FUN = slowFunction, mc.cores = nCores)))
#'
#' # Better than mc is vecorization (in this case) - no multicore required:
#' system.time(result3 <- mapply(grepl, as.list(dat$colA), as.list(dat$colB)))
#'
#' #All the same
#' all((result1$ind == result2$ind[order(result2$id)]) & result1$ind == result3)
#' @export
mclapply.0.1 <- function(X, FUN, ..., mc.preschedule = TRUE, mc.set.seed = TRUE,
                     mc.silent = FALSE, mc.cores = 1L,
                     mc.cleanup = TRUE, mc.allow.recursive = TRUE, 
                     packageToLoad = "", sourceFile = "") {
  if(.Platform$OS.type == "windows") {
    clusterFunction <- if(mc.preschedule) clusterApply else clusterApplyLB
    cl <- makeCluster(mc.cores)
    if(all(packageToLoad != "")) {
      clusterExport(cl, "packageToLoad", envir = environment())
      clusterEvalQ(cl, lapply(packageToLoad, require, character.only = TRUE))
    }
    if(sourceFile != "") {
      clusterExport(cl, "sourceFile", envir = environment())
      clusterEvalQ(cl, source(sourceFile, echo=FALSE))
    }
    result <- clusterFunction(cl = cl, x = X, fun = FUN, ...=...)
    stopCluster(cl)
    return(result)
  } else {
    return(parallel::mclapply(X = X, FUN = FUN, ...=..., mc.preschedule = mc.preschedule, 
                              mc.set.seed = mc.set.seed, mc.silent = mc.silent, mc.cores = mc.cores,
                              mc.cleanup = mc.cleanup, mc.allow.recursive = mc.allow.recursive))
  }
}